src/persist-types/src/part.rs

```diff
@@ -0,0 +1,551 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A columnar representation of one blob's worth of data
+
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use arrow2::array::{Array, NullArray, PrimitiveArray, StructArray};
+use arrow2::buffer::Buffer;
+use arrow2::chunk::Chunk;
+use arrow2::datatypes::{DataType as ArrowLogicalType, Field};
+use arrow2::io::parquet::write::Encoding;
+
+use crate::columnar::sealed::ColumnRef;
+use crate::columnar::{ColumnFormat, Data, DataType, Schema};
+
+/// A columnar representation of one blob's worth of data.
+#[derive(Debug, Default)]
+pub struct Part {
+    len: usize,
+    key: Vec<(String, DynColumnRef)>,
+    val: Vec<(String, DynColumnRef)>,
+    ts: Buffer<i64>,
+    diff: Buffer<i64>,
+}
+
+impl Part {
+    /// The number of updates contained.
+    pub fn len(&self) -> usize {
+        debug_assert_eq!(self.validate(), Ok(()));
+        self.len
+    }
+
+    /// Returns a [Columns] for the key columns.
+    pub fn key_ref<'a>(&'a self) -> ColumnsRef<'a> {
+        ColumnsRef {
+            cols: self
+                .key
+                .iter()
+                .map(|(name, col)| (name.as_str(), col))
+                .collect(),
+        }
+    }
+
+    /// Returns a [Columns] for the val columns.
+    pub fn val_ref<'a>(&'a self) -> ColumnsRef<'a> {
+        ColumnsRef {
+            cols: self
+                .val
+                .iter()
+                .map(|(name, col)| (name.as_str(), col))
+                .collect(),
+        }
+    }
+
+    pub(crate) fn to_arrow(&self) -> (Vec<Field>, Vec<Vec<Encoding>>, Chunk<Box<dyn Array>>) {
+        let (mut fields, mut encodings, mut arrays) =
+            (Vec::new(), Vec::new(), Vec::<Box<dyn Array>>::new());
+
+        {
+            let (mut key_fields, mut key_encodings, mut key_arrays) =
+                (Vec::new(), Vec::new(), Vec::new());
+            for (name, col) in self.key.iter() {
+                let (encoding, array) = col.to_arrow();
+                key_fields.push(Field::new(name, array.data_type().clone(), col.0.optional));
+                key_encodings.push(encoding);
+                key_arrays.push(array);
+            }
+            if key_arrays.is_empty() {
+                // arrow2 doesn't allow empty struct arrays
```

Comment:
none of this code handles any sort of schema migration or mismatch right now and that's on purpose for V0 (I assume you noticed that, but saying it just to be explicit). I think it's a good idea to start hashing out our story here, though!

eliding the field makes me a bit nervous[^1], but I imagine our rules could be something like this:

- no fields -> one optional field: decode expects a structarray with one column, but gets nullarray. synthesizes an structarray with a single all-null column
- N fields -> N fields + 1 optional field (where N > 0): the optional column is missing from the structarray, decode fills it in with a single all-null column
- N fields + 1 optional field -> N fields (where N >0): decode drops the relevant column from the structarray it receives
- one optional field -> no fields: decode ignores the relevant column

which all means encode is free to always write data using the current schema

thoughts?

[^1]: right now we're structuring these as logically `Vec<(K, V, T, D)>` but there's a world in which we might want to structure them as `Vec<(K, Vec<(V, Vec<(T, D)>)>)>`, which takes advantage of fancy dremel/parquet things and also more directly matches the [structure of in-memory spine batches](https://github.com/TimelyDataflow/differential-dataflow/blob/8a04699df22600a332a5fafadbe7eebb84ea16d5/src/trace/implementations/ord.rs#L88-L104). eliding things makes that more complicated

Reply:
Yeah, definitely not expecting us to have a proper schema migration story, but nice if we can avoid writing down data in such a way that it makes future schema migration stuff harder!

The n<>n+1 bullets make total sense to me. 0<>1 makes me nervous because it's the only circumstance where we'd change the type of a parquet column, which otherwise we never do. "We never change the type of a parquet column" sounds like a really nice invariant! Also "no column" to "struct column with a single field" sounds like how we'd handle schema evolution for nested structs, and I don't see a strong reason to handle it differently at the top level.

Re: the footnote... IIUC `Vec<(K, Vec<(V, Vec<(T, D)>)>)>` translates effectively to `[{k: K, vtd: [{v: V, td: [{t: T, d: D}}]}]}]` in parquet. (Just translating the tuples into records.) A null value would then be `[{k: K, vtd: [{td: [{t: T, d: D}}]}]}]`, which seems basically fine? But if you disagree this is probably something for a sync chat instead of hashing it out here... the dremel-style record shredding does get a little gnarly.

Reply:
(FWIW I don't think we necessarily need to nail this down before merging, since we're not actually persisting the results anywhere yet! It's just the one place where the correct encoding for a given schema isn't totally straightforward.)

Reply:
> 0<>1 makes me nervous because it's the only circumstance where we'd change the type of a parquet column, which otherwise we never do. "We never change the type of a parquet column" sounds like a really nice invariant!

Not sure I grokked the first time around that your point was that the column's type changed. As for the latter bit, I feel like "We always have K, V, T, and D columns" would also be a nice invariant! :D (That said, see below.)

> Also "no column" to "struct column with a single field" sounds like how we'd handle schema evolution for nested structs, and I don't see a strong reason to handle it differently at the top level.

This is a great point and actually far more compelling to me.

> `Vec<(K, Vec<(V, Vec<(T, D)>)>)>` translates effectively to `[{k: K, vtd: [{v: V, td: [{t: T, d: D}}]}]}]` in parquet. (Just translating the tuples into records.) A null value would then be `[{k: K, vtd: [{td: [{t: T, d: D}}]}]}]`, which seems basically fine? But if you disagree this is probably something for a sync chat instead of hashing it out here... the dremel-style record shredding does get a little gnarly.

👍 Yeah, good call. This also nicely parallels with DD, which special cases `(K, ())` with a different batch type that skips over the val layer.

Okay, I'm convinced! I would, however, like to hold off on any code changes for this until (a) Paul has had a chance to leave a first round of comments and (b) I've taken a moment to sketch out Struct columns as they'd be exposed to the persist user (just to confirm the intuition above about nested and top-level structs matching).

Reply:
> As for the latter bit, I feel like "We always have K, V, T, and D columns" would also be a nice invariant! :D

This is fair! I think I have a latent stress about column type changes from my time as a data infrastructure person, where it's something we'd avoid ~at all costs. (Whereas since we're writing the mapping code here, we don't necessarily have the same restrictions.)

> This also nicely parallels with DD, which special cases (K, ()) with a different batch type that skips over the val layer.

Cool!

> I would, however, like to hold off on any code changes for this until [...]

Yep, sounds good, as you prefer.